### ISSUE TITLE
Autotests are randomly failing

### ISSUE THREAD

**Issue by PeterTillema** [CE-Programming/CEmu#278]
What's wrong, and with what software version?
---------------------------------------------
**Operating System:**
Windows 10
**CEmu version:**
CEmu 1.1 (git: 5ebfd6d)
**Describe your issue:**  
The CLI autotester randomly fails on ICE autotests (https://github.com/PeterTillema/ICE/tree/parse-ast/tests); sometimes they work fine and sometimes they fail, I don't what could be the cause. It seems to me that a delay of 300ms in between pressing keys should be good enough.

What are the steps to reproduce this issue?
-------------------------------------------
1. Well, run the ICE autotests from the command line; I only saw it through Travis though (https://travis-ci.org/PeterTillema/ICE)
2. Try to test everything with the CEmu GUI as well; then it sometimes fails as well, but less often.

Any logs, error output, screenshot, other comments...?
------------------------------------------------------
Run 1:
```
Launching autotester on tests/Asm/autotester.json
tests/Asm/autotester.json loaded and verified. 2 unique tests found.
	[Test failed!] Hash #1 ("Compile program output") did not match the expected CRC A843E061 (got EC8D25CC).
	[Test failed!] Hash #2 ("Program should normally display text") did not match the expected CRC 6590D4FA (got E42D4493).
[Autotest failed] Out of 2 tests attempted, 0 passed, and 2 failed.
Launching autotester on tests/Call/autotester.json
tests/Call/autotester.json loaded and verified. 2 unique tests found.
	[Test failed!] Hash #1 ("Compile program output") did not match the expected CRC A843E061 (got EC8D25CC).
	[Test failed!] Hash #2 ("Program should call the right labels") did not match the expected CRC 3082D70A (got E42D4493).
[Autotest failed] Out of 2 tests attempted, 0 passed, and 2 failed.
Launching autotester on tests/ClrHome/autotester.json
tests/ClrHome/autotester.json loaded and verified. 2 unique tests found.
[Autotest passed] Out of 2 tests attempted, 2 passed, and 0 failed.
Launching autotester on tests/Disp/autotester.json
tests/Disp/autotester.json loaded and verified. 2 unique tests found.
	[Test failed!] Hash #1 ("Compile program output") did not match the expected CRC A843E061 (got EC8D25CC).
	[Test failed!] Hash #2 ("Program should normally display text at the right location") did not match the expected CRC 20E288BF (got E42D4493).
[Autotest failed] Out of 2 tests attempted, 0 passed, and 2 failed.
Launching autotester on tests/Goto/autotester.json
tests/Goto/autotester.json loaded and verified. 2 unique tests found.
[Autotest passed] Out of 2 tests attempted, 2 passed, and 0 failed.
Launching autotester on tests/I/autotester.json
tests/I/autotester.json loaded and verified. 2 unique tests found.
	[Test failed!] Hash #1 ("Compile program output") did not match the expected CRC A843E061 (got EC8D25CC).
	[Test failed!] Hash #2 ("Program should normally display text") did not match the expected CRC 1AE07265 (got E42D4493).
[Autotest failed] Out of 2 tests attempted, 0 passed, and 2 failed.
Launching autotester on tests/Icon/autotester.json
tests/Icon/autotester.json loaded and verified. 2 unique tests found.
[Autotest passed] Out of 2 tests attempted, 2 passed, and 0 failed.
Launching autotester on tests/Input/autotester.json
tests/Input/autotester.json loaded and verified. 3 unique tests found.
[Autotest passed] Out of 3 tests attempted, 3 passed, and 0 failed.
Launching autotester on tests/Output/autotester.json
tests/Output/autotester.json loaded and verified. 2 unique tests found.
[Autotest passed] Out of 2 tests attempted, 2 passed, and 0 failed.
```

Run 2:
```
Launching autotester on tests/Asm/autotester.json
tests/Asm/autotester.json loaded and verified. 2 unique tests found.
	[Test failed!] Hash #1 ("Compile program output") did not match the expected CRC A843E061 (got EC8D25CC).
	[Test failed!] Hash #2 ("Program should normally display text") did not match the expected CRC 6590D4FA (got E42D4493).
[Autotest failed] Out of 2 tests attempted, 0 passed, and 2 failed.
Launching autotester on tests/Call/autotester.json
tests/Call/autotester.json loaded and verified. 2 unique tests found.
	[Test failed!] Hash #1 ("Compile program output") did not match the expected CRC A843E061 (got EC8D25CC).
	[Test failed!] Hash #2 ("Program should call the right labels") did not match the expected CRC 3082D70A (got E42D4493).
[Autotest failed] Out of 2 tests attempted, 0 passed, and 2 failed.
Launching autotester on tests/ClrHome/autotester.json
tests/ClrHome/autotester.json loaded and verified. 2 unique tests found.
[Autotest passed] Out of 2 tests attempted, 2 passed, and 0 failed.
Launching autotester on tests/Disp/autotester.json
tests/Disp/autotester.json loaded and verified. 2 unique tests found.
	[Test failed!] Hash #1 ("Compile program output") did not match the expected CRC A843E061 (got EC8D25CC).
	[Test failed!] Hash #2 ("Program should normally display text at the right location") did not match the expected CRC 20E288BF (got E42D4493).
[Autotest failed] Out of 2 tests attempted, 0 passed, and 2 failed.
Launching autotester on tests/Goto/autotester.json
tests/Goto/autotester.json loaded and verified. 2 unique tests found.
[Autotest passed] Out of 2 tests attempted, 2 passed, and 0 failed.
Launching autotester on tests/I/autotester.json
tests/I/autotester.json loaded and verified. 2 unique tests found.
[Autotest passed] Out of 2 tests attempted, 2 passed, and 0 failed.
Launching autotester on tests/Icon/autotester.json
tests/Icon/autotester.json loaded and verified. 2 unique tests found.
[Autotest passed] Out of 2 tests attempted, 2 passed, and 0 failed.
Launching autotester on tests/Input/autotester.json
tests/Input/autotester.json loaded and verified. 3 unique tests found.
[Autotest passed] Out of 3 tests attempted, 3 passed, and 0 failed.
Launching autotester on tests/Output/autotester.json
tests/Output/autotester.json loaded and verified. 2 unique tests found.
	[Test failed!] Hash #1 ("Compile program output") did not match the expected CRC A843E061 (got EC8D25CC).
	[Test failed!] Hash #2 ("Program should normally display text at the right location") did not match the expected CRC C11F3369 (got CD07532D).
[Autotest failed] Out of 2 tests attempted, 0 passed, and 2 failed.
```

Run 3:
```
Launching autotester on tests/Asm/autotester.json
tests/Asm/autotester.json loaded and verified. 2 unique tests found.
[Autotest passed] Out of 2 tests attempted, 2 passed, and 0 failed.
Launching autotester on tests/Call/autotester.json
tests/Call/autotester.json loaded and verified. 2 unique tests found.
	[Test failed!] Hash #1 ("Compile program output") did not match the expected CRC A843E061 (got EC8D25CC).
	[Test failed!] Hash #2 ("Program should call the right labels") did not match the expected CRC 3082D70A (got E42D4493).
[Autotest failed] Out of 2 tests attempted, 0 passed, and 2 failed.
Launching autotester on tests/ClrHome/autotester.json
tests/ClrHome/autotester.json loaded and verified. 2 unique tests found.
	[Test failed!] Hash #1 ("Compile program output") did not match the expected CRC A843E061 (got EC8D25CC).
	[Test failed!] Hash #2 ("Program should normally display the string") did not match the expected CRC 8E38F85 (got E42D4493).
[Autotest failed] Out of 2 tests attempted, 0 passed, and 2 failed.
Launching autotester on tests/Disp/autotester.json
tests/Disp/autotester.json loaded and verified. 2 unique tests found.
	[Test failed!] Hash #1 ("Compile program output") did not match the expected CRC A843E061 (got EC8D25CC).
	[Test failed!] Hash #2 ("Program should normally display text at the right location") did not match the expected CRC 20E288BF (got E42D4493).
[Autotest failed] Out of 2 tests attempted, 0 passed, and 2 failed.
Launching autotester on tests/Goto/autotester.json
tests/Goto/autotester.json loaded and verified. 2 unique tests found.
[Autotest passed] Out of 2 tests attempted, 2 passed, and 0 failed.
Launching autotester on tests/I/autotester.json
tests/I/autotester.json loaded and verified. 2 unique tests found.
[Autotest passed] Out of 2 tests attempted, 2 passed, and 0 failed.
Launching autotester on tests/Icon/autotester.json
tests/Icon/autotester.json loaded and verified. 2 unique tests found.
[Autotest passed] Out of 2 tests attempted, 2 passed, and 0 failed.
Launching autotester on tests/Input/autotester.json
tests/Input/autotester.json loaded and verified. 3 unique tests found.
	[Test failed!] Hash #1 ("Compile program output") did not match the expected CRC A843E061 (got EC8D25CC).
	[Test failed!] Hash #3 ("Program should display proper value") did not match the expected CRC 9B2E7512 (got 2DEE525D).
[Autotest failed] Out of 3 tests attempted, 1 passed, and 2 failed.
Launching autotester on tests/Output/autotester.json
tests/Output/autotester.json loaded and verified. 2 unique tests found.
[Autotest passed] Out of 2 tests attempted, 2 passed, and 0 failed.
```

**Comment by mateoconlechuga** [comment]
Your tests just suck

**Comment by adriweb** [comment]
Yeah for now don’t rely on the autotester to do precise timing-related tests, especially if your own computer (or whereever you created the tests) is very different from the one that will run the (Travis for instance). Not that by default, SPI is enabled for the autotester. Might want to make sure this is the case in your CEmu confit as well when creating tests. 
But anyway, indeed, making the autotester single-threaded and “cycle-based” (I don’t know how to say it but I know what I mean :p), and as such, deterministic, is a known issue and should probably be somewhat prioritized... see #178 (and #275)

**Comment by adriweb** [comment]
I'll close this since it will get fixed once #178 is, so no need to track the same root issue twice.
Hopefully it'll be taken care of soon...
